test(integration): fix flaky integration tests with proper health checks

jim380 opened the pull request:
<!--
Please read and fill out this form before submitting your PR.

Please make sure you have reviewed our contributors guide before submitting your
first PR.

NOTE: PR titles should follow semantic commits: https://www.conventionalcommits.org/en/v1.0.0/
-->

## Fix Flaky Integration Tests with Proper Container Health Checks

### Problem
1. Tests were failing intermittently due to fixed sleep timers
2. JWT token handling was duplicated between test and production code
3. Resource cleanup wasn't properly ordered
4. Test organization needed improvement:
   - Multiple container setups for related tests
   - No exit status checking for subtests
   - Complex health check using configuration negotiation

### Changes

#### Test Reliability
- Replaced fixed sleep with active health checks using timer
- Optimized HTTP client and timer timeouts
- Use stateless `engine_getClientVersionV1` for health checks instead of configuration negotiation

#### Code Organization
- Added shared JWT token utility function used by both test and production code
- Moved invalid timestamp test into lifecycle test to reduce container spawns
- Added proper subtest exit status checking to fail fast on errors

#### Resource Management
- Register cleanup handlers immediately after resource creation
- Proper cleanup ordering for JWT files, Docker client, and containers
- Added `.gitkeep` to maintain jwttoken directory structure

#### File Structure
- Simplified path constants and directory handling
- Improved JWT file management in tests
- Better path handling using filepath.Join

### Testing
The changes have been tested with multiple runs of the integration tests. The tests now:
- Fail fast with clear error messages
- Clean up resources properly even when tests fail
- Run more efficiently with optimized timeouts and container reuse
- Maintain proper isolation while sharing resources where appropriate


<!-- 
Please provide an explanation of the PR, including the appropriate context,
background, goal, and rationale. If there is an issue with this information,
please provide a tl;dr and link the issue. 

Ex: Closes #15 
-->

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **Bug Fixes**
	- Improved reliability of the test setup by replacing fixed sleep duration with dynamic polling for Reth container readiness.

- **New Features**
	- Introduced a new function for generating a dynamic JWT secret, enhancing security during integration tests.
	- Added a new function to encapsulate JWT token generation, simplifying error handling and improving code organization.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jim380 committed:
test(integration): add health check

jim380 committed:
test(integration): generate elastic jwt secret each run

ProgramCpp commented on integration_test.go:
```diff
@@ -106,21 +127,82 @@ func setupTestRethEngine(t *testing.T) {
+	err = waitForRethContainer(t, jwtSecret)
+	require.NoError(t, err)
 
 	t.Cleanup(func() {
```
register each cleanup step separately, after each stateful operation so that any subsequent error leaves the test env clean

ProgramCpp commented on integration_test.go:
```diff
@@ -106,21 +127,82 @@ func setupTestRethEngine(t *testing.T) {
+					}
+				}
+			}
+			time.Sleep(500 * time.Millisecond)
```
the container is typically ready in 100 milliseconds in the local env. can reduce the poll interval. earlier, the tests failed with wait time of 50 milliseconds

tzdybal replied:
I agree that 100ms or even 50ms pooling time will be OK.

ProgramCpp commented on integration_test.go:
```diff
@@ -206,7 +288,7 @@ func TestExecutionClientLifecycle(t *testing.T) {
 }
 
 func TestExecutionClient_InitChain_InvalidPayloadTimestamp(t *testing.T) {
```
optimization: can move this test into a subtest of the above lifecycle test, to avoid spinning up the reth containers for multiple tests

ProgramCpp commented on integration_test.go:
```diff
@@ -106,21 +127,82 @@ func setupTestRethEngine(t *testing.T) {
+					}
+					req.Header.Set("Content-Type", "application/json")
+
+					jwtSecretBytes, err := hex.DecodeString(strings.TrimPrefix(jwtSecret, "0x"))
```
extract the auth bearer token logic into a utility that can be both used by the test and the actual code.
```

func getAuthToken(jwtSecret string) (string, error) {
	if jwtSecret != "" {
		secret, err := hex.DecodeString(jwtSecret)
		if err != nil {
			return "", err
		}

		token := jwt.NewWithClaims(jwt.SigningMethodHS256, jwt.MapClaims{
			"exp": time.Now().Add(time.Hour * 1).Unix(), // Expires in 1 hour
			"iat": time.Now().Unix(),
		})

		// Sign the token with the decoded secret
		authToken, err := token.SignedString(secret)
		if err != nil {
			return "", err
		}
		return authToken, nil
	}
	return "", nil
}
```

ProgramCpp commented on integration_test.go:
```diff
@@ -106,21 +127,82 @@ func setupTestRethEngine(t *testing.T) {
+func waitForRethContainer(t *testing.T, jwtSecret string) error {
+	t.Helper()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
```
this context is never used. and cancel is redundant. can we use a timer instead?

ProgramCpp replied:
cancel is redundant - the main routine is blocked anyway

MSevey replied:
agreed, since the context is only being used as a timer, we should just use a timer.

ProgramCpp commented on integration_test.go:
```diff
@@ -106,21 +127,82 @@ func setupTestRethEngine(t *testing.T) {
+	defer cancel()
+
+	client := &http.Client{
+		Timeout: 1 * time.Second,
```
1 second timeout is too high for a local network call

ProgramCpp commented on integration_test.go:
```diff
@@ -106,21 +127,82 @@ func setupTestRethEngine(t *testing.T) {
+				resp.Body.Close()
+				if resp.StatusCode == http.StatusOK {
+					// check :8551 is ready
+					req, err := http.NewRequest("POST", TEST_ENGINE_URL, strings.NewReader(`{"jsonrpc":"2.0","method":"engine_exchangeTransitionConfigurationV1","params":[],"id":1}`))
```
can we keep the configurations and negotiations between the clients in the initChain call only. for tests, a stateless call like engine_getClientVersionV1 can be used?

ProgramCpp commented on integration_test.go:
```diff
@@ -206,7 +288,7 @@ func TestExecutionClientLifecycle(t *testing.T) {
 }
```
check exit status of sub tests, to terminate the test on failure. 

```
require.True(t, t.Run("InitChain", func(t *testing.T) {
...
}))
```

MSevey reviewed (changes requested):
agree with @ProgramCpp review comments

jim380 committed:
test: address PR#16 comments

- Replace fixed sleep with active health checks using timer
- Use stateless engine_getClientVersionV1 for health checks
- Move JWT token generation to shared utility function
- Register cleanup handlers immediately after resource creation
- Check subtest exit status to fail fast
- Move invalid timestamp test into lifecycle test
- Optimize timeouts for faster test execution

jim380 committed:
ci: fix lint

tzdybal reviewed (approved):
LGTM

tzdybal commented:
Created follow-up issue: https://github.com/rollkit/go-execution-evm/issues/18

MSevey reviewed (approved):
utACK, defer to @tzdybal
